src/LoginHandler.php

```diff
@@ -0,0 +1,261 @@
+<?php
+namespace SilverStripe\MFA;
+
+use LogicException;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\MFA\Extensions\MemberExtension;
+use SilverStripe\MFA\Model\AuthenticationMethod;
+use SilverStripe\Security\Member;
+use SilverStripe\Security\MemberAuthenticator\LoginHandler as BaseLoginHandler;
+use SilverStripe\Security\MemberAuthenticator\MemberLoginForm;
+
+class LoginHandler extends BaseLoginHandler
+{
+    const SESSION_KEY = 'MFALogin';
+
+    private static $url_handlers = [
+        'mfa/start/$Method' => 'start',
+        'mfa/verify' => 'verify',
+        'mfa' => 'mfa',
+    ];
+
+    private static $allowed_actions = [
+        'mfa',
+        'startMethod',
+    ];
+
+    /**
+     * Indicate how many MFA methods the user must authenticate with before they are considered logged in
+     *
+     * @config
+     * @var int
+     */
+    private static $required_mfa_methods = 1;
+
+    /**
+     * A "session store" object that helps contain MFA specific session detail
+     *
+     * @var SessionStore
+     */
+    protected $sessionStore;
+
+    /**
+     * Override the parent "doLogin" to insert extra steps into the flow
+     *
+     * @inheritdoc
+     */
+    public function doLogin($data, MemberLoginForm $form, HTTPRequest $request)
+    {
+        $member = $this->checkLogin($data, $request, $result);
+
+        // If there's no member it's an invalid login. We'll delegate this to the parent
+        if (!$member) {
+            return parent::doLogin($data, $form, $request);
+        }
+
+        // Store a reference to the member in session
+        $this->getSessionStore()->setMember($member);
+        $this->getSessionStore()->save($request);
+
+        // Store the BackURL for use after the process is complete
+        if (!empty($data)) {
+            $request->getSession()->set(static::SESSION_KEY . '.additionalData', $data);
+        }
+
+        // Redirect to the MFA step
+        return $this->redirect($this->link('mfa'));
+    }
+
+    /**
+     * Action handler for displaying the MFA authentication React app
+     *
+     * @return array|HTTPResponse
+     */
+    public function mfa()
+    {
+        $member = $this->getSessionStore()->getMember();
+
+        // If we don't have a valid member we shouldn't be here...
+        if (!$member) {
+            return $this->redirectBack();
+        }
+
+        // Get a list of authentication for the user and the find default
+        $authMethods = $member->AuthenticationMethods();
+
+        // Pool a list of "lead in" labels. We skip the default here assuming it's not required.
+        $alternateLeadInLabels = [];
```

Comment:
I think there's potential here for something (currently missing) that could help with the concept of "given I know which authenticators are registered, let me do something with their instances". What about an `AuthenticatorInterfaceRegistry` service which loads registered authenticators for either the current member, or from config (all available authenticators) and maybe exposes a callable method of some sort like `AuthenticatorInterfaceRegistry::aggregate(callable $callback)` so you could do something like `$methods = AuthenticatorInterfaceRegistry::aggregate(function (AuthenticatorInterface $authenticator) { return ...});`

I might be over-engineering this, but I think it might be useful if you want to get a map of callbacks for each authenticator's start() method for example. Or maybe I'm overthinking it.

If you don't like this idea, maybe we move this particular block of code into its own function somewhere instead to keep it separated.

Reply:
I considered writing some service class for collecting MFA method classes. It will probably still prove useful when registration is built in. Currently there's no configuration for enabled methods but we will need it. 

Happy to get that ball rolling in this PR if that makes sense